src/plugins/socketmon/socketmon.cpp

```diff
@@ -1379,6 +1382,287 @@ static event_response_t udpb_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
 
 /* ----------------------------------------------------- */
 
+static void print_dns_info(drakvuf_t drakvuf, drakvuf_trap_info_t* info, socketmon* sm, const char* function_name, const char* dns_name)
+{
+    switch (sm->format)
+    {
+        case OUTPUT_CSV:
+            printf("socketmon," FORMAT_TIMEVAL ",%" PRIu32 ",0x%" PRIx64 ",\"%s\",\"%s\",\"%s\"\n",
+                   UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3,
+                   info->proc_data.name, dns_name, function_name);
+            break;
+
+        case OUTPUT_KV:
+            printf("socketmon Time=" FORMAT_TIMEVAL ",PID=%d,PPID=%d,ProcessName=\"%s\",DnsName=\"%s\",Method=\"%s\"\n",
+                   UNPACK_TIMEVAL(info->timestamp), info->proc_data.pid, info->proc_data.ppid,
+                   info->proc_data.name, dns_name, function_name);
+            break;
+
+        default:
+        case OUTPUT_DEFAULT:
+            printf("[SOCKETMON] TIME:" FORMAT_TIMEVAL " VCPU:%" PRIu32 " CR3:0x%" PRIx64 ",\"%s\" DNS_NAME:\"%s\" METHOD:\"%s\"\n",
+                   UNPACK_TIMEVAL(info->timestamp), info->vcpu, info->regs->cr3, info->proc_data.name,
+                   dns_name, function_name);
+            break;
+    };
+}
+
+static event_response_t trap_DnsQuery_A_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    socketmon* sm = (socketmon*)info->trap->data;
+
+    addr_t domain_name_addr = 0;
+
+    if (sm->pm == VMI_PM_IA32E) // only 64 bit for now
+    {
+        domain_name_addr = info->regs->rcx;
+
+        access_context_t ctx;
+        ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+        ctx.dtb = info->regs->cr3;
+        ctx.addr = domain_name_addr;
+
+        char* dns_name = [&]
+        {
+            vmi_lock_guard vmi_lg(drakvuf);
+            return vmi_read_str(vmi_lg.vmi, &ctx);
+        }();
+        print_dns_info(drakvuf, info, sm, info->trap->name, dns_name);
+        g_free(dns_name);
+    }
+    else
+    {
+        fprintf(stderr, "[SOCKETMON] Suddenly, we are in 32 bit mode in %s(...) trap. Unsupported.\n", info->trap->name);
+    }
+
+    return 0;
+}
+
+
+static event_response_t trap_DnsQuery_W_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    socketmon* sm = (socketmon*)info->trap->data;
+
+    addr_t domain_name_addr = 0;
+    unicode_string_t* domain_name_us = nullptr;
+
+    if (sm->pm == VMI_PM_IA32E) // only 64 bit for now
+    {
+        domain_name_addr = info->regs->rcx;
+
+        access_context_t ctx;
+        ctx.translate_mechanism = VMI_TM_PROCESS_DTB;
+        ctx.dtb = info->regs->cr3;
+        ctx.addr = domain_name_addr;
+
+        {
```

Comment:
Why these braces and why do you need vmi_lock_guard here?

Reply:
These braces denote a scope. It is used to employ an idiom called RAII (Resource Acquisition Is Initialization).

RAII is used to safely access things like files, mutexes or any other resource, that must be freed before function exit.
One creates a variable on stack of a lock type, here "vmi_lock_guard", that locks on construction. And on destruction it frees the resource. C++ guarantees to call destructor of stack variable on scope exit.

Here is more info about it:
https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization

Reply:
That is pretty cool but my worry about advanced C++ features like this is they tend to extremely slow. Even constructors tend to be extremely slow. So using these functions in event callbacks is problematic.

Reply:
It might be worthwile performance testing this as well to determine how much overhead it adds call a constructor/destructor for every event callback. In my experience doing high-volume data processing they tend to be major bottlenecks.